Frameworks/Foundation/NSNotificationCenter.mm

```diff
@@ -14,227 +14,289 @@
+        for (NSString* key in copiedAllKeys) {
+            StrongId<NSMutableSet<_NSNotificationReceiver*>> nameReceivers = [_observers objectForKey:key];
+            [nameReceivers minusSet:receivers];
+            if ([nameReceivers count] == 0) {
+                [_observers removeObjectForKey:name];
+            }
+        }
+        [copiedAllKeys release];
     }
-    IwFree(sendTo);
 }
 
 /**
  @Status Interoperable
 */
-- (void)postNotificationName:(NSString*)name object:(NSObject*)sender userInfo:(NSDictionary*)userInfo {
-    NSNotification* notification = [NSNotification notificationWithName:name object:sender userInfo:userInfo];
+- (void)postNotificationName:(NSString*)name object:(NSObject*)sender {
+    [self postNotificationName:name object:sender userInfo:nil];
+}
 
+/**
+ @Status Interoperable
+*/
+- (void)postNotificationName:(NSString*)name object:(NSObject*)sender userInfo:(NSDictionary*)userInfo {
+    // Since object could be mid-deallocation, this notification must not be leaked into an autorelease pool.
+    StrongId<NSNotification> notification;
+    notification.attach([[NSNotification alloc] initWithName:name object:sender userInfo:userInfo]);
     [self postNotification:notification];
 }
 
 /**
  @Status Interoperable
 */
-- (void)addObserver:(id)observer selector:(SEL)selName name:(NSString*)name object:(id)object {
-    if (name == nil) {
+- (void)postNotification:(NSNotification*)notification {
+    NSString* name = [notification name];
+
+    // This method does not use _receiversForName:(...) above as it treats a nil _notification_ sender as a wildcard.
+    // We want to treat a nil _receiver_ sender as a wildcard.
+    std::vector<StrongId<_NSNotificationReceiver>> validReceivers{ 0 };
+    @synchronized(self) {
+        NSMutableSet<_NSNotificationReceiver*>* receivers = [self _receiverSetForNotificationName:name createIfMissing:NO];
+        if (!receivers) {
+            return;
+        }
+
+        NSObject* sender = [notification object];
+        validReceivers.reserve([receivers count]);
+        for (_NSNotificationReceiver* receiver in receivers) {
+            if (receiver->sender == nil || receiver->sender == sender) {
+                validReceivers.emplace_back(receiver);
+            }
+        }
+    }
+
+    // Notification post happens outside of the lock over self; receivers are allowed to
+    // manipulate the notification center's state.
+
+    // Receivers are retained for the duration of this notification post.
+    for (auto& receiver : validReceivers) {
+        [receiver postNotification:notification];
+    }
+}
+
+- (void)_addObserver:(id)observer selector:(SEL)selector name:(NSString*)name object:(id)object withOwningReference:(BOOL)owningReference {
+    if (!name || !observer || !selector) {
         return;
     }
 
-    NSNotificationReceiver* newObserver = [NSNotificationReceiver new];
+    _NSNotificationReceiver* newReceiver = [_NSNotificationReceiver new];
```

Comment:
I am very sad that StrongId requires a two-step init for consuming a +1 reference.